### ISSUE TITLE
Compile Warnings "sign-compare" in covercrosses and coverbins

### ISSUE THREAD

**Issue by MarkTakaWilliams** [amiq-consulting/fc4sc#3]
Does not let you compile with -Werror enabled due to 

```
fc4sc-master/includes/fc4sc_bin.hpp:356:23: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
         if (bin_index == count)
                       ^
```

If bin_index was converted to size_t this warning could be fixed, this issue also exists in the covercross code.

**Comment by amiq-consulting** [comment]
Hi,

This issue should already be resolved in the latest commit on branch master (https://github.com/amiq-consulting/fc4sc/commit/5db2b4f89a1b378f7363181e66c832ab8b34c28d). 
Note that this commit also adds a change which forces the user to provide a name to bin instantiation, so the following expression will not compile anymore:
`bin<int>(1,2);`
Instead, you will have to use:
`bin<int>("some_name",1,2);`

Best Regards

**Comment by amiq-consulting** [comment]
Hi,

We have made a FC4SC release containing new functionality and changes that add to the quality of the library. These updates also contain the fixes to the following issues: #2 #3 #5
See details below.

### New Coverpoint definition method
## New
```c++
int data_value = 0;
// The coverpoint will samples the value returned by the 
// evaluation of the "data_value * 2" expression.
// The evaluation is done at the call to sample() function
// of the parent covergroup
COVERPOINT(int, data_cvp, data_value * 2) {
    illegal_bin<int>("illegal_zero", 0),
    bin<int>("positive", interval(1,INT_MAX)),
    bin<int>("negative", interval(-1,INT_MIN))
};
```
## Old
```c++
// Step 2: bind variable to coverpoint
int SAMPLE_POINT(data_value, data_cvp);
// Step 1:  declare coverpoint
coverpoint<int> data_cvp = coverpoint<int> (
    this,
    illegal_bin<int>("illegal_zero", 0),
    bin<int>("positive", interval(1,INT_MAX)),
    bin<int>("negative", interval(-1,INT_MIN))
);
// Step 3: update coverpoint variable and sample it 
void sample(int data) {
    this->data_value = data;
    covergroup::sample();
}
```
## Benefits for the new syntax:
1. Gets rid of the extra step of using SAMPLE_POINT(...)) macro which has the following problems:
    - A _**segmentation fault**_ would be generated if SAMPLE_POINT(...); was declared after its associated coverpoint is instantiated (Note above: Step 2 precedes Step 1)
    - Relied on the user manually binding a sample variable to the coverpoint, which is error prone
    - Restricted sampling to be made on the value of a variable, not supporting complex/compound expressions
2. Adds support for **sample expressions** & **sample conditions**
3. Custom covergroup sample function is no longer required
4. It is much clearer, easier to use and much more similar to SystemVerilog standard coverpoint definitions.

_**The old definition syntax is still (currently) supported, but considered deprecated and will be removed in the future!**_

### New Coverpoint definition restriction
Coverpoints must contain at least one bin; Empty coverpoints are no longer allowed!
Note: this constraint is valid for the new definition method only!

### New Bin definition restrictions
   - Bin name must now be specified; Unnamed bins are no longer allowed!
       - _Note: The name must be the first argument to the constructor!_
   - The bin must contain at least one value/interval; Empty bins are no longer allowed!

Example:
```c++
auto my_bin = bin<int>(
    "power_of_2",
    1,              // {1}
    interval(2,3),  // [2:3]
    interval(7,4),  // [4:7]
    interval(15,8)  // [8:15]
);
```
_**If any of the coverpoint/bin definition restrictions are not met, compilation will fail!**_

All these updates can be found in the github repository, here: [FC4SC_v2.0](https://github.com/amiq-consulting/fc4sc/tree/FC4SC_v2.0)
Any feedback is appreciated.

Best Regards
